x-pack/plugins/elastic_assistant/server/lib/langchain/execute_custom_llm_chain/index.ts

```diff
@@ -5,47 +5,70 @@
  * 2.0.
  */
 
-import { KibanaRequest } from '@kbn/core/server';
+import { ElasticsearchClient, KibanaRequest, Logger } from '@kbn/core/server';
 import type { PluginStartContract as ActionsPluginStart } from '@kbn/actions-plugin/server';
-import { ConversationChain } from 'langchain/chains';
 import { BufferMemory, ChatMessageHistory } from 'langchain/memory';
 import { BaseMessage } from 'langchain/schema';
 
-import { ActionsClientLlm } from '../llm/actions_client_llm';
+import { ChatPromptTemplate, SystemMessagePromptTemplate } from 'langchain/prompts';
+import { ConversationalRetrievalQAChain } from 'langchain/chains';
 import { ResponseBody } from '../helpers';
+import { ActionsClientLlm } from '../llm/actions_client_llm';
+import { ElasticsearchStore } from '../elasticsearch_store/elasticsearch_store';
+import { KNOWLEDGE_BASE_INDEX_PATTERN } from '../../../routes/knowledge_base/constants';
 
 export const executeCustomLlmChain = async ({
   actions,
   connectorId,
+  esClient,
   langChainMessages,
+  logger,
   request,
 }: {
   actions: ActionsPluginStart;
   connectorId: string;
+  esClient: ElasticsearchClient;
   langChainMessages: BaseMessage[];
+  logger: Logger;
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   request: KibanaRequest<unknown, unknown, any, any>;
 }): Promise<ResponseBody> => {
   const llm = new ActionsClientLlm({ actions, connectorId, request });
 
+  // Chat History Memory: in-memory memory, from client local storage, first message is the system prompt
   const pastMessages = langChainMessages.slice(0, -1); // all but the last message
   const latestMessage = langChainMessages.slice(-1); // the last message
-
+  const systemPrompt = langChainMessages[0].content; // the first message
   const memory = new BufferMemory({
     chatHistory: new ChatMessageHistory(pastMessages),
+    memoryKey: 'chat_history',
   });
 
-  const chain = new ConversationChain({ llm, memory });
+  // ELSER backed ElasticsearchStore for Knowledge Base
+  const esStore = new ElasticsearchStore(esClient, KNOWLEDGE_BASE_INDEX_PATTERN, logger);
 
-  await chain.call({ input: latestMessage[0].content }); // kick off the chain with the last message
+  const prompt = ChatPromptTemplate.fromPromptMessages([
+    SystemMessagePromptTemplate.fromTemplate(systemPrompt),
+  ]);
+
+  // Chain w/ chat history memory and knowledge base retriever
+  const chain = ConversationalRetrievalQAChain.fromLLM(llm, esStore.asRetriever(), {
+    memory,
+    qaChainOptions: { prompt, type: 'stuff' },
+  });
+  await chain.call({ question: latestMessage[0].content });
+
+  // Chain w/ just knowledge base retriever
+  // const chain = RetrievalQAChain.fromLLM(llm, esStore.asRetriever(), {
+  //   prompt,
+  // });
+  // await chain.call({ query: latestMessage[0].content });
```

Comment:
thanks for keeping this here one more round, as we're still experementing with different chains

Reply:
`prompt` removed from example as well 👍